source/_posts/2015-03-26-creating-a-dynamic-single-page-app-for-our-genome-team-using-react.markdown

```diff
@@ -0,0 +1,297 @@
+
+#### From CoffeeScript to React: Selecting Artworks
+
+In this app, genomers are able to 'select' artworks on the panel for the purposes of saving and conducting batch actions. In our initial implementation, clicking the 'select all' button would individually click each artwork and used global event listeners to change UI state:
+
+```coffee
+($ 'body').on 'click', '.artwork-selector', (evt) ->
+  container = $(this)
+  currentSlug = container.data('id')
+  artworkIdsElement = $('#selected_artwork_ids')
+  selectedArtworkIds = _.compact(_.uniq(artworkIdsElement.val().split(',')))
+  indexOfCurrentSlug = selectedArtworkIds.indexOf(currentSlug)
+  if selectedArtworkIds.indexOf(currentSlug) != -1
+    selectedArtworkIds.splice(indexOfCurrentSlug, 1)
+    container.text('Not selected')
+    container.removeClass('btn-purple')
+  else
+    selectedArtworkIds.push currentSlug
+    container.text('Selected')
+    container.addClass('btn-purple')
+  $('#selected_artwork_ids').val(selectedArtworkIds.join(','))
+  return false
+```
+
+With React, we store whether or not an artwork is selected as part of our state, and the appearance of elements results from this variable. We use [class sets](https://facebook.github.io/react/docs/class-name-manipulation.html) to dynamically alter styles such as button color. When the `selected` state changes, React re-renders all components that depend on that variable.
+
+```javascript
+var SelectedButton = React.createClass({
+  changeIsSelected: function (e) {
+    e.preventDefault();
+    var newSelectedState = !this.props.artwork.isSelected;
+    this.props.updateArtwork(this.props.artwork._id, function(artwork) {
+      artwork.isSelected = newSelectedState;
+      return artwork;
+    })
+  },
+  render: function() {
+    var cx = React.addons.classSet;
+    var selectedButtonClasses = cx({
+      'btn-purple': this.props.artwork.isSelected,
+      'btn-tiny': true,
+      'btn': true,
+      'artwork-selector': true
+    });
+    return (
+      <div className="panel-artwork-actions">
+        <a className={selectedButtonClasses}
+           data-id='false'
+           href='#'
+           onClick={ this.changeIsSelected }>
+         {this.props.artwork.isSelected ? '' : 'Not '}Selected</a>
+      </div>
+    )
+  }
+});
+```
+
+### Challenges
+
+#### React's Virtual DOM
+React keeps track of a Virtual DOM created by components you define. This can lead to issues, especially when trying to integrate React with jQuery plugins. For example, our modals kept showing up within other components until we explicitly rendered them on the outermost level. We also had issues trying to use an existing drag/drop plugin with the way we set up our state, and ended up [building one from scratch](https://gist.github.com/sweir27/4ea941dd717da69527d6).
```

Comment:
i know we value transparency at artsy, but the `building one from scratch` links to a bit of code with a comment in it that says: `// add a comment for what this clever line does?` haha

so maybe in this case we should not link :)